osquery/core/database/rocksdb_database.h

```diff
@@ -0,0 +1,106 @@
+  ~RocksdbDatabase() override {}
+
+  ExpectedSuccess<DatabaseError> destroyDB() override;
+  ExpectedSuccess<DatabaseError> open() override;
+
+  // Open and close are not thread safe commands,
+  // Before closing db you need to ensure that no other theads are currenly
+  // using db connection.
+  // Please see comment above about reopening db.
+  void close() override;
+
+  Expected<int32_t, DatabaseError> getInt32(const std::string& domain,
+                                            const std::string& key) override;
+  Expected<std::string, DatabaseError> getString(
+      const std::string& domain, const std::string& key) override;
+
+  ExpectedSuccess<DatabaseError> putInt32(const std::string& domain,
+                                          const std::string& key,
+                                          const int32_t value) override;
+  ExpectedSuccess<DatabaseError> putString(const std::string& domain,
+                                           const std::string& key,
+                                           const std::string& value) override;
+
+  Expected<std::vector<std::string>, DatabaseError> getKeys(
+      const std::string& domain, const std::string& prefix = "") override;
+
+  // This function should be used only as optimization
+  // This write operation will not use neither sync or WAL, so date lose
+  // may happen in case of failure, but opertaion itself is still atomic
+  ExpectedSuccess<DatabaseError> putStringsUnsafe(
+      const std::string& domain,
+      std::vector<std::pair<std::string, std::string>>& data) override;
+
+ private:
+  rocksdb::Options getOptions();
+  std::vector<rocksdb::ColumnFamilyDescriptor> createDefaultColumnFamilies(
+      const rocksdb::Options& options);
+  ExpectedSuccess<DatabaseError> openInternal(
+      const rocksdb::Options& options, const boost::filesystem::path& path);
+  Expected<std::string, DatabaseError> getRawBytesInternal(
+      Handle* handle, const std::string& key);
+  ExpectedSuccess<DatabaseError> putRawBytesInternal(Handle* handle,
+                                                     const std::string& key,
+                                                     const std::string& value);
+  ExpectedSuccess<DatabaseError> checkDbConnection();
+  Expected<std::shared_ptr<Handle>, DatabaseError> getHandle(
+      const std::string& domain);
+
+  Expected<std::string, DatabaseError> getRawBytes(const std::string& domain,
+                                                   const std::string& key);
+  ExpectedSuccess<DatabaseError> putRawBytes(const std::string& domain,
+                                             const std::string& key,
+                                             const std::string& value);
+
+ private:
+  bool in_panic_ = false;
+  rocksdb::ReadOptions default_read_options_;
+  rocksdb::WriteOptions default_write_options_;
+  rocksdb::WriteOptions batch_write_options_;
+  std::unique_ptr<rocksdb::DB> db_ = nullptr;
+
+  std::string path_;
```

Comment:
1. AFAIK `path_` never changes during the lifetime of object, maybe it's better to define it as `const`.
2. Would you like to use for it more specific type `boost::filesystem::path` ?